managed/services/uievents/uievents.go

```diff
@@ -0,0 +1,262 @@
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package uievents provides facility to store UI events.
+package uievents
+
+import (
+	"context"
+	"encoding/json"
+	"sort"
+	"time"
+
+	"github.com/HdrHistogram/hdrhistogram-go"
+	pmmv1 "github.com/percona-platform/saas/gen/telemetry/events/pmm"
+	"github.com/sirupsen/logrus"
+
+	uievents "github.com/percona/pmm/api/uieventspb"
+	"github.com/percona/pmm/managed/services/telemetry"
+)
+
+const (
+	cleanupAfterHours    = 30
+	cleanupCheckInterval = 1 * time.Minute
+)
+
+// Service provides facility for storing UI events.
+type Service struct {
+	l               *logrus.Entry
+	lastCleanup     time.Time
+	dashboardUsage  map[string]*DashboardUsageStat
+	componentsUsage map[string]*ComponentsUsageStat
+
+	uievents.UnimplementedUIEventsServer
+}
+
+type DashboardUsageStat struct {
+	title    string
+	uid      string
+	useCount int32
+	loadTime *hdrhistogram.Histogram
+}
+
+type ComponentsUsageStat struct {
+	uid      string
+	useCount int32
+	loadTime *hdrhistogram.Histogram
+}
+
+// New returns platform Service.
+func New() (*Service, error) {
+	l := logrus.WithField("component", "platform")
+
+	s := Service{
+		l:               l,
+		dashboardUsage:  make(map[string]*DashboardUsageStat),
+		componentsUsage: make(map[string]*ComponentsUsageStat),
+		lastCleanup:     time.Now(),
+	}
+
+	return &s, nil
+}
+
+// ScheduleCleanup schedules internal clean for internal UI event storage.
+func (s *Service) ScheduleCleanup(ctx context.Context) {
+	go func() {
+		ticker := time.NewTicker(cleanupCheckInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				s.cleanupIfNeeded()
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+func (s *Service) FetchMetrics(ctx context.Context, report *telemetry.Config) ([]*pmmv1.ServerMetric_Metric, error) {
+	var metrics []*pmmv1.ServerMetric_Metric
+
+	metrics = s.processDashboardStat(metrics)
+	metrics = s.processComponentsStat(metrics)
+
+	return metrics, nil
+}
+
+func (s *Service) processDashboardStat(metrics []*pmmv1.ServerMetric_Metric) []*pmmv1.ServerMetric_Metric {
+	type Stat struct {
+		TopDashboards         []string `json:"top_dashboards"`
+		SlowDashboardsP95_1s  []string `json:"slow_dashboards_p95_1s"`
+		SlowDashboardsP95_5s  []string `json:"slow_dashboards_p95_5s"`
+		SlowDashboardsP95_10s []string `json:"slow_dashboards_p95_10s"`
+	}
```

Comment:
can we extract this struct outside of the method?

Reply:
The point here is visibility, this type makes sense only in this function.
You think it has performance implications?
There is none:

```

import (
	"testing"
)

type StatOutside struct {
	TopDashboards         []string `json:"top_dashboards"`
	SlowDashboardsP95_1s  []string `json:"slow_dashboards_p95_1s"`
	SlowDashboardsP95_5s  []string `json:"slow_dashboards_p95_5s"`
	SlowDashboardsP95_10s []string `json:"slow_dashboards_p95_10s"`
}

func Benchmark(b *testing.B) {
	b.Run("outside", func(b *testing.B) {
		items := make([]StatOutside, 0, 100)
		for i := 0; i < 100; i++ {
			items = append(items, StatOutside{})

		}
	})
	b.Run("inside", func(b *testing.B) {
		type StatInside struct {
			TopDashboards         []string `json:"top_dashboards"`
			SlowDashboardsP95_1s  []string `json:"slow_dashboards_p95_1s"`
			SlowDashboardsP95_5s  []string `json:"slow_dashboards_p95_5s"`
			SlowDashboardsP95_10s []string `json:"slow_dashboards_p95_10s"`
		}

		items := make([]StatInside, 0, 100)
		for i := 0; i < 100; i++ {
			items = append(items, StatInside{})

		}
	})
}
```

running with `go test -bench=. -benchtime=10000000x -count=10`

```
goos: darwin
goarch: arm64
pkg: github.com/percona/pmm/agent/agents/mysql/slowlog/parser
Benchmark/outside-10            10000000                 0.0005083 ns/op
Benchmark/outside-10            10000000                 0.0004250 ns/op
Benchmark/outside-10            10000000                 0.0005834 ns/op
Benchmark/outside-10            10000000                 0.0004292 ns/op
Benchmark/outside-10            10000000                 0.0003334 ns/op
Benchmark/outside-10            10000000                 0.0003917 ns/op
Benchmark/outside-10            10000000                 0.0006208 ns/op
Benchmark/outside-10            10000000                 0.0005125 ns/op
Benchmark/outside-10            10000000                 0.0002625 ns/op
Benchmark/outside-10            10000000                 0.0003083 ns/op
Benchmark/inside-10             10000000                 0.0004750 ns/op
Benchmark/inside-10             10000000                 0.0002750 ns/op
Benchmark/inside-10             10000000                 0.0005667 ns/op
Benchmark/inside-10             10000000                 0.0003541 ns/op
Benchmark/inside-10             10000000                 0.0002542 ns/op
Benchmark/inside-10             10000000                 0.0004916 ns/op
Benchmark/inside-10             10000000                 0.0002750 ns/op
Benchmark/inside-10             10000000                 0.0002625 ns/op
Benchmark/inside-10             10000000                 0.0002542 ns/op
Benchmark/inside-10             10000000                 0.0003375 ns/op
```